File: Dynamic programming - algos/lis.cpp
/////////////////////////Problem Statement/////////////////////////

///////////////////

The Longest Increasing Subsequence (LIS) problem is to find the length of the longest subsequence of a given sequence such that all elements of the subsequence are sorted in increasing order. For example, the length of LIS for {10, 22, 9, 33, 21, 50, 41, 60, 80} is 6 and LIS is {10, 22, 33, 50, 60, 80}.
longest-increasing-subsequence

Given a sequence, find the length of the longest increasing subsequence from a given sequence .
The longest increasing subsequence means to find a subsequence of a given sequence in which the subsequence's elements are in sorted order, lowest to highest, and in which the subsequence is as long as possible. This subsequence is not necessarily contiguous, or unique.

Note: Duplicate numbers are not counted as increasing subsequence.

For example:
 length of LIS for 
{ 10, 22, 9, 33, 21, 50, 41, 60, 80 } is 6 and LIS is {10, 22, 33, 50, 60, 80}.

Input:

The first line contains an integer T, depicting total number of test cases. 
Then following T lines contains an integer N depicting the size of array and next line followed by the value of array.


Output:

Print the Max length of the subsequence in a separate line.


Constraints:

1 ≤ T ≤ 100
1 ≤ N ≤ 1000
0 ≤ A[i] ≤ 300

///////////////////






////////////////////////////Solution///////////////////////////////////


#include <bits/stdc++.h>
using namespace std;

int lis(vector<int> array)
{
    vector<int> dp(array.size(),1);
    int n=array.size(),max_val=0;
             for (int i=0;i<n-1;i++)
             {
                for (int j=i+1;j<n;j++)
                {
                               if(array[i]<array[j])
                               {
                                if(dp[j]<dp[i]+1)
                                {
                                    dp[j]=dp[i]+1;
                                }
                               }                
                }
    }
    for (int i=0;i<n;i++)
    {
        max_val=max(dp[i],max_val);
    }
    return max_val;
}


int main() 
{
    int T;
    cin>>T;
    int n;
    while(T--)
    {
        cin>>n;
        vector<int>  array(n,0);
        for(int i=0;i<n;i++)
        {
                cin>>array[i];
        }
        cout<<lis(array)<<endl;
    }
    return 0;
}

File: Dynamic programming - algos/LIS.java
import java.util.Scanner;

public class LIS{
	/* In this problem we have to find length of longest increasing subsequence such that
	 * all elements of a subsequence are sorted in increasing order.
	 * for example : {40, 10, 20, 15, 50, 80} LIS is 4
	 *               i.e {10,20,50,80}
	 */
	public static void main(String[] args) {
		Scanner s= new Scanner(System.in);
		
		System.out.println("Please enter the size of array");
		int n= s.nextInt();
		// declaring n sized array..
		int arr[] = new int[n];
		// taking array input..
		for(int i=1;i<=n;i++){
			System.out.println("Please enter element "+i +":");
			arr[i-1]= s.nextInt();		
		}
		// lenLongestIncSubs() will return length of increasing subsequence..
		System.out.println(lenLongestIncSubs(arr,n));
		s.close();
	}

	public static int lenLongestIncSubs(int[] arr, int n) {
		int dp[] = new int[n];		// this will store the result for each array index
									// i.e. length of inc subseq. till that index considering it the last element of array...
		int max = 1;				
		for(int i=0;i<n;i++){
		   // initializing all elements result with 1 considering a single element is always in inc seq.
		   dp[i] =1;
		   
		   /* here we will have to go to each element before the current one
		    *to check if it is smaller than current one.. and if it is 
		    *then we will add its result to the current one..
		    *and we will keep on checking further and in last result at ith index
		    *will be updated with max len of inc subs. till itself..
		    */
		   for(int j=0;j<i;j++){
			   // checking if jth element is smaller than ith so as to form inc subs.
				if(arr[j]<arr[i]){
					// we will only update the res when new res is greater than previous..
					if(dp[i]<dp[j]+1){
						dp[i] = dp[j]+1;
					}
					// updating our max if it becomes less than current result at ith element...
					if(max<dp[i])
						max = dp[i];
				}
			}
		}
		return max;
	}

}
